third_party/terraform/utils/retry_transport_test.go

```diff
@@ -0,0 +1,275 @@
+		func(w http.ResponseWriter, r *http.Request) {
+			w.WriteHeader(400)
+			if _, err := w.Write([]byte(testRetryTransportErrorMessageFailure)); err != nil {
+				t.Errorf("unable to write to response writer: %s", err)
+			}
+		}))
+	defer ts.Close()
+
+	client := ts.Client()
+	client.Transport = &retryTransport{
+		internal:        http.DefaultTransport,
+		retryPredicates: []RetryErrorPredicateFunc{testRetryTransportRetryPredicate},
+	}
+
+	resp, err := client.Get(ts.URL)
+	testRetryTransportCheckResponseForFailure(t, resp, err, 400, testRetryTransportErrorMessageFailure)
+}
+
+// Check for no errors if the request succeeds after a certain amount of time
+func TestRetryTransport_SuccessAfterRetries(t *testing.T) {
+	ts := httptest.NewServer(testRetryTransportSucceedAfterHandler(t, time.Second*1))
+	defer ts.Close()
+
+	client := ts.Client()
+	client.Transport = &retryTransport{
+		internal:        http.DefaultTransport,
+		retryPredicates: []RetryErrorPredicateFunc{testRetryTransportRetryPredicate},
+	}
+
+	ctx, cc := context.WithTimeout(context.Background(), time.Second*2)
+	defer cc()
+	req, err := http.NewRequestWithContext(ctx, "GET", ts.URL, nil)
+	if err != nil {
+		t.Fatalf("unable to construct err: %v", err)
+	}
+
+	resp, err := client.Do(req)
+	testRetryTransportCheckResponseForSuccess(t, resp, err)
+}
+
+func TestRetryTransport_FailAfterRetries(t *testing.T) {
+	ts := httptest.NewServer(testRetryTransportFailAfterHandler(t, time.Second*1))
+	defer ts.Close()
+
+	client := ts.Client()
+	client.Transport = &retryTransport{
+		internal:        http.DefaultTransport,
+		retryPredicates: []RetryErrorPredicateFunc{testRetryTransportRetryPredicate},
+	}
+
+	ctx, cc := context.WithTimeout(context.Background(), time.Second*2)
+	defer cc()
+	req, err := http.NewRequestWithContext(ctx, "GET", ts.URL, nil)
+	if err != nil {
+		t.Fatalf("unable to construct err: %v", err)
+	}
+
+	resp, err := client.Do(req)
+	testRetryTransportCheckResponseForFailure(t, resp, err, 400, testRetryTransportErrorMessageFailure)
+}
+
+func TestRetryTransport_ContextTimeout(t *testing.T) {
+	ts := httptest.NewServer(testRetryTransportSucceedAfterHandler(t, time.Second*4))
+	defer ts.Close()
+
+	client := ts.Client()
+	client.Transport = &retryTransport{
+		internal:        http.DefaultTransport,
+		retryPredicates: []RetryErrorPredicateFunc{testRetryTransportRetryPredicate},
+	}
+
+	ctx, cc := context.WithTimeout(context.Background(), time.Second*2)
+	defer cc()
+	req, err := http.NewRequestWithContext(ctx, "GET", ts.URL, nil)
+	if err != nil {
+		t.Fatalf("unable to construct err: %v", err)
+	}
+	resp, err := client.Do(req)
+	// Last failure should have been a retryable error since we timed out
+	testRetryTransportCheckResponseForFailure(t, resp, err, 500, testRetryTransportErrorMessageRetry)
+}
+
+// Check for no errors if the request succeeds after a certain amount of time
+func TestRetryTransport_SuccessAfterRetriesWithBody(t *testing.T) {
```

Comment:
What does this test compared to the others? That our request body is making it through, or something else?

Reply:
yup, edited a bit though so outdated